opentracing-vertx-web/src/main/java/io/opentracing/contrib/vertx/ext/web/TracingHandler.java

```diff
@@ -0,0 +1,118 @@
+package io.opentracing.contrib.vertx.ext.web;
+
+import io.opentracing.contrib.vertx.ext.web.WebSpanDecorator.StandardTags;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import io.opentracing.Span;
+import io.opentracing.SpanContext;
+import io.opentracing.Tracer;
+import io.opentracing.propagation.Format;
+import io.opentracing.tag.Tags;
+import io.vertx.core.Handler;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import io.vertx.ext.web.RoutingContext;
+
+/**
+ * Handler which creates tracing data for all server requests. It should be added to
+ * {@link io.vertx.ext.web.Route#handler(Handler)} and {@link io.vertx.ext.web.Route#failureHandler(Handler)} as the
+ * first in the chain.
+ *
+ * @author Pavol Loffay
+ */
+public class TracingHandler implements Handler<RoutingContext> {
+    private static final Logger log = LoggerFactory.getLogger(TracingHandler.class);
+    public static final String CURRENT_SPAN = TracingHandler.class.getName() + ".severSpan";
+
+    private final Tracer tracer;
+    private final List<WebSpanDecorator> decorators;
+
+    public TracingHandler(Tracer tracer) {
+        this(tracer, Collections.singletonList(new StandardTags()));
+    }
+
+    public TracingHandler(Tracer tracer, List<WebSpanDecorator> decorators) {
+        this.tracer = tracer;
+        this.decorators = new ArrayList<>(decorators);
+    }
+
+    @Override
+    public void handle(RoutingContext routingContext) {
+        if (routingContext.failed()) {
+            handlerFailure(routingContext);
+        } else {
+            handlerNormal(routingContext);
+        }
+    }
+
+    protected void handlerNormal(RoutingContext routingContext) {
+        // reroute
+        Object object = routingContext.get(CURRENT_SPAN);
+        if (object != null) {
+            Span span = (Span) object;
+            decorators.forEach(spanDecorator ->
+                    spanDecorator.onReroute(routingContext.request(), span));
+
+            // TODO in 3.3.3 it was sufficient to add this when creating the span
+            routingContext.addBodyEndHandler(finishEndHandler(routingContext, span));
```

Comment:
@pavolloffay no, bodyend handler is a routing context event that is triggered once a response has been written, say e.g.: the `end` has been called. You should be careful since if there are errors/response is not sent, then this method is not called. So I think if you want to do any action in the `finishEndHandler` code it is not guaranteed that it will be called.